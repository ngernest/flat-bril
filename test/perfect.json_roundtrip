{
  "functions": [
    {
      "args": [
        {
          "name": "input",
          "type": "int"
        }
      ],
      "instrs": [
        {
          "args": [
            "input"
          ],
          "dest": "n",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "dest": "v0",
          "op": "const",
          "type": "int",
          "value": 0
        },
        {
          "dest": "v1",
          "op": "const",
          "type": "int",
          "value": 1
        },
        {
          "dest": "v2",
          "op": "const",
          "type": "int",
          "value": 2
        },
        {
          "args": [
            "v1"
          ],
          "dest": "sum",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v2"
          ],
          "dest": "i",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "args": [
            "v1"
          ],
          "dest": "result",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "label": "for.cond.1"
        },
        {
          "args": [
            "i",
            "i"
          ],
          "dest": "ii",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "ii",
            "n"
          ],
          "dest": "comp",
          "funcs": [],
          "labels": [],
          "op": "gt",
          "type": "bool"
        },
        {
          "args": [
            "comp"
          ],
          "funcs": [],
          "labels": [
            "for.end",
            "for.body"
          ],
          "op": "br"
        },
        {
          "label": "for.body"
        },
        {
          "args": [
            "n",
            "i"
          ],
          "dest": "qut",
          "funcs": [],
          "labels": [],
          "op": "div",
          "type": "int"
        },
        {
          "args": [
            "qut",
            "i"
          ],
          "dest": "mpt",
          "funcs": [],
          "labels": [],
          "op": "mul",
          "type": "int"
        },
        {
          "args": [
            "n",
            "mpt"
          ],
          "dest": "dif",
          "funcs": [],
          "labels": [],
          "op": "sub",
          "type": "int"
        },
        {
          "args": [
            "dif",
            "v0"
          ],
          "dest": "comp1",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "comp1"
          ],
          "funcs": [],
          "labels": [
            "if.body",
            "for.incre"
          ],
          "op": "br"
        },
        {
          "label": "if.body"
        },
        {
          "args": [
            "sum",
            "i"
          ],
          "dest": "sum",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [
            "sum",
            "qut"
          ],
          "dest": "sum",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "label": "for.incre"
        },
        {
          "args": [
            "i",
            "v1"
          ],
          "dest": "i",
          "funcs": [],
          "labels": [],
          "op": "add",
          "type": "int"
        },
        {
          "args": [],
          "funcs": [],
          "labels": [
            "for.cond.1"
          ],
          "op": "jmp"
        },
        {
          "label": "for.end"
        },
        {
          "args": [
            "sum",
            "n"
          ],
          "dest": "comp2",
          "funcs": [],
          "labels": [],
          "op": "eq",
          "type": "bool"
        },
        {
          "args": [
            "comp2"
          ],
          "funcs": [],
          "labels": [
            "if.success",
            "if.failure.end"
          ],
          "op": "br"
        },
        {
          "label": "if.success"
        },
        {
          "args": [
            "v0"
          ],
          "dest": "result",
          "funcs": [],
          "labels": [],
          "op": "id",
          "type": "int"
        },
        {
          "label": "if.failure.end"
        },
        {
          "args": [
            "result"
          ],
          "funcs": [],
          "labels": [],
          "op": "print"
        }
      ],
      "name": "main"
    }
  ]
}
