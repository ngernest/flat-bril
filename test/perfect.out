@main
	op: id   		dest: n 		type: int		args: ["input"]	
	op: const		dest: v0		type: int		value: 0	
	op: const		dest: v1		type: int		value: 1	
	op: const		dest: v2		type: int		value: 2	
	op: id   		dest: sum		type: int		args: ["v1"]	
	op: id   		dest: i 		type: int		args: ["v2"]	
	op: id   		dest: result		type: int		args: ["v1"]	
.for.cond.1
	op: mul  		dest: ii		type: int		args: ["i", "i"]	
	op: gt   		dest: comp		type: bool		args: ["ii", "n"]	
	op: br   		args: ["comp"]		labels: ["for.end", "for.body"]
.for.body
	op: div  		dest: qut		type: int		args: ["n", "i"]	
	op: mul  		dest: mpt		type: int		args: ["qut", "i"]	
	op: sub  		dest: dif		type: int		args: ["n", "mpt"]	
	op: eq   		dest: comp1		type: bool		args: ["dif", "v0"]	
	op: br   		args: ["comp1"]		labels: ["if.body", "for.incre"]
.if.body
	op: add  		dest: sum		type: int		args: ["sum", "i"]	
	op: add  		dest: sum		type: int		args: ["sum", "qut"]	
.for.incre
	op: add  		dest: i 		type: int		args: ["i", "v1"]	
	op: jmp  		labels: ["for.cond.1"]
.for.end
	op: eq   		dest: comp2		type: bool		args: ["sum", "n"]	
	op: br   		args: ["comp2"]		labels: ["if.success", "if.failure.end"]
.if.success
	op: id   		dest: result		type: int		args: ["v0"]	
.if.failure.end
	op: print		args: ["result"]	
